templates/cloudflare-workers/worker/bookmarkUnfurling.ts

```diff
@@ -0,0 +1,88 @@
+import { IRequest, error } from 'itty-router'
+
+export async function handleUnfurlUrl(request: IRequest) {
+	const url = request.query.url
+	if (typeof url !== 'string' || !url.startsWith('http://') || !url.startsWith('https://')) {
+		return error(400, 'Invalid URL')
+	}
+
+	// cloudflare has a built-in HTML parser/rewriter called HTMLRewriter. in order to use it, we
+	// need to define classes that act as event handlers for certain elements, attributes, etc.
+	// see https://developers.cloudflare.com/workers/runtime-apis/html-rewriter/
+	const meta$ = new MetaExtractor()
+	const title$ = new TextExtractor()
+	const icon$ = new IconExtractor()
+
+	try {
+		await new HTMLRewriter()
+			.on('meta', meta$)
+			.on('title', title$)
+			.on('link', icon$)
+			.transform((await fetch(url)) as any)
+			.blob()
+	} catch {
+		return null
+	}
```

Comment:
(followup work) I want us to put this whole thing in a npm package. Maybe even in a separate repo under MIT. It's definitely an unfulfilled need, and people will probably make PRs with improvements for different edge cases.